File: app/src/main/java/me/wcy/music/service/MediaSessionManager.java
package me.wcy.music.service;

import android.os.Build;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;

import me.wcy.music.application.AppCache;
import me.wcy.music.model.Music;
import me.wcy.music.utils.CoverLoader;

/**
 * Created by hzwangchenyan on 2017/8/8.
 */
public class MediaSessionManager {
    private static final String TAG = "MediaSessionManager";
    private static final long MEDIA_SESSION_ACTIONS = PlaybackStateCompat.ACTION_PLAY
            | PlaybackStateCompat.ACTION_PAUSE
            | PlaybackStateCompat.ACTION_PLAY_PAUSE
            | PlaybackStateCompat.ACTION_SKIP_TO_NEXT
            | PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS
            | PlaybackStateCompat.ACTION_STOP
            | PlaybackStateCompat.ACTION_SEEK_TO;

    private PlayService playService;
    private MediaSessionCompat mediaSession;

    public static MediaSessionManager get() {
        return SingletonHolder.instance;
    }

    private static class SingletonHolder {
        private static MediaSessionManager instance = new MediaSessionManager();
    }

    private MediaSessionManager() {
    }

    public void init(PlayService playService) {
        this.playService = playService;
        setupMediaSession();
    }

    private void setupMediaSession() {
        mediaSession = new MediaSessionCompat(playService, TAG);
        mediaSession.setFlags(MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS | MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS);
        mediaSession.setCallback(callback);
        mediaSession.setActive(true);
    }

    public void updatePlaybackState() {
        int state = (AudioPlayer.get().isPlaying() || AudioPlayer.get().isPreparing()) ? PlaybackStateCompat.STATE_PLAYING : PlaybackStateCompat.STATE_PAUSED;
        mediaSession.setPlaybackState(
                new PlaybackStateCompat.Builder()
                        .setActions(MEDIA_SESSION_ACTIONS)
                        .setState(state, AudioPlayer.get().getAudioPosition(), 1)
                        .build());
    }

    public void updateMetaData(Music music) {
        if (music == null) {
            mediaSession.setMetadata(null);
            return;
        }

        MediaMetadataCompat.Builder metaData = new MediaMetadataCompat.Builder()
                .putString(MediaMetadataCompat.METADATA_KEY_TITLE, music.getTitle())
                .putString(MediaMetadataCompat.METADATA_KEY_ARTIST, music.getArtist())
                .putString(MediaMetadataCompat.METADATA_KEY_ALBUM, music.getAlbum())
                .putString(MediaMetadataCompat.METADATA_KEY_ALBUM_ARTIST, music.getArtist())
                .putLong(MediaMetadataCompat.METADATA_KEY_DURATION, music.getDuration())
                .putBitmap(MediaMetadataCompat.METADATA_KEY_ALBUM_ART, CoverLoader.get().loadThumb(music));

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            metaData.putLong(MediaMetadataCompat.METADATA_KEY_NUM_TRACKS, AppCache.get().getLocalMusicList().size());
        }

        mediaSession.setMetadata(metaData.build());
    }

    private MediaSessionCompat.Callback callback = new MediaSessionCompat.Callback() {
        @Override
        public void onPlay() {
            AudioPlayer.get().playPause();
        }

        @Override
        public void onPause() {
            AudioPlayer.get().playPause();
        }

        @Override
        public void onSkipToNext() {
            AudioPlayer.get().next();
        }

        @Override
        public void onSkipToPrevious() {
            AudioPlayer.get().prev();
        }

        @Override
        public void onStop() {
            AudioPlayer.get().stopPlayer();
        }

        @Override
        public void onSeekTo(long pos) {
            AudioPlayer.get().seekTo((int) pos);
        }
    };
}

File: app/src/main/java/me/wcy/music/service/MediaSessionManager.kt
package me.wcy.music.service

import android.content.Context
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import me.wcy.music.storage.db.entity.SongEntity
import top.wangchenyan.common.CommonApp
import top.wangchenyan.common.utils.image.ImageUtils

/**
 * Created by hzwangchenyan on 2017/8/8.
 */
class MediaSessionManager(
    private val context: Context,
    private val audioPlayer: AudioPlayer
) {
    private val mediaSession: MediaSessionCompat by lazy {
        MediaSessionCompat(context, TAG).apply {
            setFlags(MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS or MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS)
            setCallback(callback)
            isActive = true
        }
    }
    private var loadCoverJob: Job? = null

    fun updatePlaybackState() {
        val state = if (audioPlayer.playState.value.isPlaying
            || audioPlayer.playState.value.isPreparing
        ) {
            PlaybackStateCompat.STATE_PLAYING
        } else {
            PlaybackStateCompat.STATE_PAUSED
        }
        mediaSession.setPlaybackState(
            PlaybackStateCompat.Builder()
                .setActions(MEDIA_SESSION_ACTIONS)
                .setState(state, audioPlayer.getAudioPosition(), 1f)
                .build()
        )
    }

    fun updateMetaData(song: SongEntity?) {
        loadCoverJob?.cancel()
        if (song == null) {
            mediaSession.setMetadata(null)
            return
        } else {
            val builder = MediaMetadataCompat.Builder()
                .putString(MediaMetadataCompat.METADATA_KEY_TITLE, song.title)
                .putString(MediaMetadataCompat.METADATA_KEY_ARTIST, song.artist)
                .putString(MediaMetadataCompat.METADATA_KEY_ALBUM, song.album)
                .putString(MediaMetadataCompat.METADATA_KEY_ALBUM_ARTIST, song.artist)
                .putLong(MediaMetadataCompat.METADATA_KEY_DURATION, song.duration)
                .putLong(
                    MediaMetadataCompat.METADATA_KEY_NUM_TRACKS,
                    (audioPlayer.playlist.value?.size ?: 0).toLong()
                )
            mediaSession.setMetadata(builder.build())
            loadCoverJob = CommonApp.appScope.launch {
                val bitmap = ImageUtils.loadBitmap(song.getLargeCover()).data
                if (bitmap != null) {
                    builder.putBitmap(
                        MediaMetadataCompat.METADATA_KEY_ALBUM_ART,
                        bitmap
                    )
                    mediaSession.setMetadata(builder.build())
                }
            }
        }
    }

    fun getMediaSession(): MediaSessionCompat.Token {
        return mediaSession.sessionToken
    }

    private val callback: MediaSessionCompat.Callback = object : MediaSessionCompat.Callback() {
        override fun onPlay() {
            audioPlayer.startPlayer()
        }

        override fun onPause() {
            audioPlayer.pausePlayer()
        }

        override fun onSkipToNext() {
            audioPlayer.next()
        }

        override fun onSkipToPrevious() {
            audioPlayer.prev()
        }

        override fun onStop() {
            audioPlayer.stopPlayer()
        }

        override fun onSeekTo(pos: Long) {
            audioPlayer.seekTo(pos.toInt())
        }
    }

    companion object {
        private const val TAG = "MediaSessionManager"
        private const val MEDIA_SESSION_ACTIONS = (PlaybackStateCompat.ACTION_PLAY
                or PlaybackStateCompat.ACTION_PAUSE
                or PlaybackStateCompat.ACTION_PLAY_PAUSE
                or PlaybackStateCompat.ACTION_SKIP_TO_NEXT
                or PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS
                or PlaybackStateCompat.ACTION_STOP
                or PlaybackStateCompat.ACTION_SEEK_TO)
    }
}